CSI Volume clone

brybacki opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
https://github.com/kubevirt/containerized-data-importer/issues/1467

In short this is a rewrite of PR: "Feat CSI Clone Support " https://github.com/kubevirt/containerized-data-importer/pull/1552 to utilize "Cluster scoped DataVolume/PVC namespace transfer API" #1673 

**Special notes for your reviewer**:

PR is is now open for review but is not yet in the good shape. The main design is here, and I see at least 3 things to be done:
- [x] Correct the clone/expansion coordination - there is a bug somewhere, I am now working on it
- [x] Extract common code  - there is big similarity between CSI Volume Clone and Smart Clone, some code can be made common/simplified
- [x] Documentation


**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
CDI can utilize CSI Volume Cloning
```

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.17-ceph

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.19-ceph

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.20-hpp

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.19-ceph

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.19-hpp

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.19-ceph

brybacki commented:
/test all

brybacki commented:
PR is is now open for review but is not yet in the good shape. The main design is here, and I see at least 3 things to be done:
- Correct the clone/expansion coordination - there is a bug somewhere, I am now working on it
- Extract common code  - there is big similarity between CSI Volume Clone and Smart Clone, some code can be made common/simplified
- Documentation

brybacki commented:
/retest

awels reviewed (commented):
So I am a little confused [according to this](https://kubernetes.io/docs/concepts/storage/volume-pvc-datasource/) all we have to do is create a PVC with a datasource. Which would eliminate most if not all of the csi clone controller. In the DV controller if we decide that we are going to csi clone, just make the PVC and set the datasource to be the source defined in the DV. What am I missing here?

brybacki commented:
> So I am a little confused [according to this](https://kubernetes.io/docs/concepts/storage/volume-pvc-datasource/) all we have to do is create a PVC with a datasource. Which would eliminate most if not all of the csi clone controller. In the DV controller if we decide that we are going to csi clone, just make the PVC and set the datasource to be the source defined in the DV. What am I missing here?

You are right, I must have been blinded by fighting some bugs for a few days ;). I have simplified the controllers by removing the CSI - clone controller, and added one check to the DV controller.

awels commented on pkg/controller/datavolume-controller.go:
```diff
@@ -595,24 +553,194 @@ func (r *DatavolumeReconciler) reconcileSmartClonePvc(log logr.Logger, datavolum
+		}
+
+		if pvc.Status.Phase == corev1.ClaimBound {
+			if v, ok := pvc.Annotations[AnnCloneOf]; !ok || v != "true" {
```
I see this code several times, maybe we can make it a function, like 'setCloneOfOnPvc'?

brybacki replied:
Done

awels commented on pkg/controller/datavolume-controller.go:
```diff
@@ -595,24 +553,194 @@ func (r *DatavolumeReconciler) reconcileSmartClonePvc(log logr.Logger, datavolum
+	}
+
+	// trigger transfer and next reconcile should have pvcExists == true
+	tmpPVC.Annotations[annReadyForTransfer] = "true"
```
expandTmpPvcAfterClone and expandPvcAfterClone look very similar, except for some additional annotations and the updateCloneStatusPhase is different. I think maybe we can combine some of that?

brybacki replied:
I did try extracting the expand:
```
func (r *DatavolumeReconciler) expandPvcAfterClone(log logr.Logger,
	datavolume *cdiv1.DataVolume,
	pvc *corev1.PersistentVolumeClaim,
	pvcSpec *corev1.PersistentVolumeClaimSpec,
	selectedCloneStrategy cloneStrategy) (bool, error) {
	done, err := r.expand(log, datavolume, pvc, pvcSpec)
	if err != nil {
		return false, err
	}
	if !done {
		return false, r.updateCloneStatusPhase(cdiv1.ExpansionInProgress, datavolume, pvc, selectedCloneStrategy)
	}

	// done
	return true, nil
}
```

But it does not look much better in my opinion.  Any ideas?

awels commented on pkg/controller/datavolume-controller.go:
```diff
@@ -1483,6 +1605,59 @@ func newSnapshot(dataVolume *cdiv1.DataVolume, snapshotName, snapshotClassName s
 }
 
+// NewVolumeClonePVC creates a PVC object to be used during CSI volume cloning. csiClonePvcType defines whether the PVC object will be a source cloner PVC or the target PVC
+func newVolumeClonePVC(dv *cdiv1.DataVolume, sourcePvc *corev1.PersistentVolumeClaim, targetPvcSpec *corev1.PersistentVolumeClaimSpec, pvcName string) *corev1.PersistentVolumeClaim {
```
Can we just call newPersistentVolumeClaim, and override the pvc.Spec.DataSource before returning the result. That way all the logic that sets annotations and ownerRefs and everything else is contained in that function.

brybacki replied:
I can try, we will see if that is better

brybacki replied:
Done

mhenriks commented on pkg/controller/datavolume-controller.go:
```diff
@@ -595,24 +546,201 @@ func (r *DatavolumeReconciler) reconcileSmartClonePvc(log logr.Logger, datavolum
+		return reconcile.Result{}, err
+	}
+	if inUse || !populated {
+		return reconcile.Result{Requeue: true},
```
This is an okay use of requeue because we may not be watching/reconciling these pods

mhenriks commented on pkg/controller/datavolume-controller.go:
```diff
@@ -595,24 +546,201 @@ func (r *DatavolumeReconciler) reconcileSmartClonePvc(log logr.Logger, datavolum
+
+	//DO Nothing, not yet ready
+	if pvc.Annotations[AnnCloneOf] != "true" {
+		return reconcile.Result{Requeue: true}, nil
```
Shouldn't have to requeue

brybacki replied:
Removed

mhenriks commented on pkg/controller/datavolume-controller.go:
```diff
@@ -595,24 +546,201 @@ func (r *DatavolumeReconciler) reconcileSmartClonePvc(log logr.Logger, datavolum
+		}
+	}
+
+	return reconcile.Result{Requeue: true}, r.updateCloneStatusPhase(cdiv1.CSICloneInProgress, datavolume, nil, CsiClone)
```
This requeue necessary?

brybacki replied:
I was under impression, that when we do Clone, we might want to requeue the same way as in Import to check progress/update status, but no I think that for CSI Clone it does not make any sense, there is nothing to reconcile until PVC is either Bound or Lost (or has some warning event)

Removed

brybacki replied:
Looks like it does not work without requeue here, investigating.

brybacki replied:
Some coordination between Namespace transfer controller and datavolume controller is probably not handled explicitely in this case, when I add Requeue it works, but I think I can be implement in a better way.

brybacki replied:
finally got it, not needed, there was a wrong owner was set on transfer/temporary pvc (a DV from different namespace), and it was being deleted. I am not sure if k8s GC was deleting it or our logic in some other branch of code, but now after fixing that problem it works.

mhenriks reviewed (commented):
Looks pretty good but agree with @awels re: duplicated code.  Do what you can to clean up but would like to get this in.

See comments about requeue.  Usually only be necessary when waiting on resources that are not watched to change

brybacki commented on pkg/controller/datavolume-controller.go:
```diff
@@ -425,92 +433,30 @@ func (r *DatavolumeReconciler) Reconcile(_ context.Context, req reconcile.Reques
+				}
+			} else if pvc.Status.Phase == corev1.ClaimPending {
+				//not yet ready
+				return reconcile.Result{Requeue: true}, nil
```
right, and the Phase will change

brybacki commented on pkg/controller/datavolume-controller.go:
```diff
@@ -425,92 +433,30 @@ func (r *DatavolumeReconciler) Reconcile(_ context.Context, req reconcile.Reques
+				//not yet ready
+				return reconcile.Result{Requeue: true}, nil
+			}
+			// TODO: is there any work for ClaimLost,
```
Do you think it is ok to do:
```
phase = cdiv1.Failed, 
eventType: corev1.EventTypeWarning,
reason:    ErrClaimLost,
```

When you see ClaimLost, then it is probably better to investigate, delete DV (to cleanup any transfer objects etc...), fix what is wrong and retry.

mhenriks replied:
I've seen cases become unlost so as long as "failed" is not terminal

awels replied:
I can in fact give an example of becoming 'unlost'. Create a manual PV and PVC, and ensure they become bound. Then 'export' the yaml for both. Then delete the PV and PVC. Then create both PV and PVC again, the PVC will be lost for a little bit until it gets matched back with the PV.

brybacki replied:
ok, so do you suggest a new phase - like  CloneTargetPVCLost, or leaving the Failed and just testing the scenario to make sure it can eventually continue if PVC changed from ClaimLost to Bound?

brybacki replied:
Updated one condition and according to new test the CSI clone logic can continue when PVC is recovered from ClaimLost

brybacki replied:
https://github.com/kubevirt/containerized-data-importer/blob/b04a47574e3e58a322d22bfc5629b55a2bfbc314/pkg/controller/datavolume-controller_test.go#L1195

brybacki commented:
/retest

awels commented:
/retest

brybacki commented:
documentation updated

brybacki commented:
@awels @mhenriks please take a look, it is now ready

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.20-upg

awels commented on tests/utils/storageprofile.go:
```diff
@@ -0,0 +1,75 @@
+	clientSet *cdiclientset.Clientset,
+	storageClassName string,
+	spec *cdiv1.StorageProfileSpec,
+	cloneStrategyCsiClone cdiv1.CDICloneStrategy) error {
```
this is for configuring any clone strategy right? Why is the variable called cloneStrategyCsiClone?

brybacki commented on pkg/controller/datavolume-controller.go:
```diff
@@ -600,24 +557,201 @@ func (r *DatavolumeReconciler) reconcileSmartClonePvc(log logr.Logger, datavolum
+			r.updateCloneStatusPhase(cdiv1.CloneScheduled, datavolume, nil, CsiClone)
+	}
+
+	// here use Temp PVC!!!
```
It was an old TODO comment to myself, removed now

brybacki commented on tests/utils/storageprofile.go:
```diff
@@ -0,0 +1,75 @@
+		return nil, err
+	}
+
+	originalProfileSpec := storageProfile.Spec.DeepCopy()
```
No

brybacki commented on pkg/controller/datavolume-controller.go:
```diff
@@ -600,24 +557,201 @@ func (r *DatavolumeReconciler) reconcileSmartClonePvc(log logr.Logger, datavolum
+		return reconcile.Result{}, err
+	}
+	pvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: cloneTargetPvc.Namespace, Name: cloneTargetPvc.Name}, pvc); err != nil {
```
this makes sense, I'll do it

brybacki replied:
Now I think I'll leave it like it is now:

Currently it does:
get:
- notExists -> create
- exists -> check for name clash

New way would be like this, and probably harder to cover all possibilities:
create:
- created - ok
- alreadyExists:
 -- get:
    -exists -> check for name clash
    - *not exists (low probability but actually might happen)*

@awels  What do you think

awels replied:
I was trying to figure out why you would need to get the PVC again, since you have everything you need from the PVC definition you attempted to created, but then I noticed in the else you are checking the pvc.Status.Phase, which you can only get from querying the PVC. For this reason I am okay with leaving it as is.

brybacki commented on pkg/controller/datavolume-controller.go:
```diff
@@ -600,24 +557,201 @@ func (r *DatavolumeReconciler) reconcileSmartClonePvc(log logr.Logger, datavolum
+
+	pvcName := datavolume.Name
+
+	if isCrossNamespaceClone(datavolume) {
```
doCrossNamespaceClone done

brybacki commented:
I have applied almost all the changes suggested in the comments. Only one change request was not applied:

https://github.com/kubevirt/containerized-data-importer/pull/1824/files?file-filters%5B%5D=.bazel&file-filters%5B%5D=.go&file-filters%5B%5D=.md#diff-8f140feb34d5514d8336c8181a67578e3d4938a6bade5129cf24d78d80986f45R587

"So why not simply attempt to create the PVC, and if that fails with an IsAlreadyExists drop into the else part. Something like this:"

brybacki commented:
/retest

awels reviewed (approved):
/lgtm

mhenriks commented:
/lgtm
/approve

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.19-ceph

awels commented:
/override pull-containerized-data-importer-e2e-k8s-1.19-ceph
